components/brave_rewards/resources/extension/brave_rewards/background/publisherVisit.ts

```diff
@@ -0,0 +1,227 @@
+        type: 'MediaDurationMetadataRequest',
+        url: details.url
+      })
+    },
+    // Filters
+    {
+      types: [
+        'image',
+        'media',
+        'script',
+        'xmlhttprequest'
+      ],
+      urls: [
+        data.urlRegex
+      ]
+    })
+}
+
+const getPublisherPanelInfo = (tabId: number, publisherKey: string) => {
+  chrome.braveRewards.getPublisherPanelInfo(
+    publisherKey, (result: RewardsExtension.Result, info?: RewardsExtension.Publisher) => {
+      if (result === 0 && info) {
+        rewardsPanelActions.onPublisherData(tabId, info)
+      }
+      return
+    })
+}
+
+const savePublisherInfo = (tabId: number, mediaType: string, url: string, publisherKey: string, publisherName: string, favIconUrl: string) => {
+  chrome.braveRewards.savePublisherInfo(
+    tabId,
+    mediaType,
+    url,
+    publisherKey,
+    publisherName,
+    favIconUrl,
+    (result: RewardsExtension.Result) => {
+      if (result !== 0) {
+        console.error(`Failed to save publisher info for ${publisherKey}, result is ${result}`)
+        return
+      }
+    })
+}
+
+const handleSavePublisherVisitResponse = (tabId: number, mediaType: string, data: SavePublisherVisitResponse) => {
+  if (!data.publisherKey) {
+    console.error('Failed to handle publisher visit: missing publisher key')
+    return
+  }
+
+  if (data.mediaKey && !publisherKeys[data.mediaKey]) {
+    publisherKeys[data.mediaKey] = data.publisherKey
+  }
+
+  chrome.braveRewards.getPublisherInfo(
+    data.publisherKey, (result: RewardsExtension.Result, info?: RewardsExtension.Publisher) => {
+      if (result === 0 && info) {
+        getPublisherPanelInfo(tabId, data.publisherKey)
+        return
+      }
+
+      // Failed to find publisher info corresponding to this key, so save it now
+      if (result === 9) {
+        savePublisherInfo(
+          tabId,
+          mediaType,
+          data.url,
+          data.publisherKey,
+          data.publisherName,
+          data.favIconUrl || '')
+        return
+      }
+    })
+}
+
+chrome.runtime.onMessageExternal.addListener((msg, sender, sendResponse) => {
+  if (!sender || !sender.tab || !msg) {
+    return
+  }
+
+  const windowId = sender.tab.windowId
+  if (!windowId) {
+    return
+  }
+
+  const tabId = sender.tab.id
+  if (!tabId) {
+    return
+  }
+
+  const response = msg as GreaselionResponse
+  if (!response.data) {
+    console.error(`Received empty Greaselion response payload for ${msg.type} message`)
+    return
+  }
+
+  switch (msg.type) {
+    case 'GreaselionError': {
+      const data = response.data as GreaselionErrorResponse
+      handleGreaselionErrorResponse(tabId, response.mediaType, data)
+      break
+    }
+    case 'SavePublisherVisit': {
+      const data = response.data as SavePublisherVisitResponse
+      handleSavePublisherVisitResponse(tabId, response.mediaType, data)
+      break
+    }
+  }
+})
+
+chrome.runtime.onConnectExternal.addListener(function (port) {
```

Comment:
similar to https://github.com/brave/brave-core/pull/5440/files#r475844579: can we validate port.sender.id?

Reply:
Same issue as above, unfortunately.